### PR TITLE
Permutation extended

### PR DESCRIPTION
**Pull request by jjsjann123**
Extended permutation support in integration (See more details on #1601). This update allows us to better support permutation propagation on tensors, specifically for binary ops with inputs of different ranks. Our goal is to avoid permuting tensors unless absolutely necessary. We try to preserve the permutation propagation rule in aten, with some known limitation at the time.

The idea in this implementation is the same as with our existing code, which is to permute input/output tensors outside of codegen: For a simplified binary op scenario: `output = binaryOp(input0, input1)`

1. In a simple case where `input0` and `input1` come with the same rank & permutation order, our output would preserve the same permutation;
2. For cases where `input0` and `input1` come with different ranks but with **compatible** permutation, the tensor with the higher rank dictates the permutation of the output;
3. For cases where `input0` and `input1` come with different ranks but with **in-compatible** permutation, this is where permutation propagation fails and the output tensor will be contiguous.

By **compatible** permutation, it means that we can permute the higher rank tensor to contiguous format, and then apply a second permutation to the tensor with lower rank to match their axes. This check is implemented in `MemoryFormat::broadcastToRank(int lower_rank)`.

Some concrete example (note that we comply with eager propagation on cases 1-3, but diverge in behavior for cases 4, 5):
1. different rank & same permutation
```
    t0 = torch.randn(b, h, w, c).cuda().permute([0, 3, 1, 2])  # stride (hwc, 1, wc, c)
    t1 = torch.randn(h, w, c).cuda().permute([2, 0, 1])  # stride (1, wc, c)
    out = scripted_add(t0, t1)  # stride (hwc, 1, wc, c) preserving memory format of t0
```
2. different rank & compatible permutation
```
    t0 = torch.randn(b, h, w, c).cuda().permute([0, 3, 1, 2])  # stride (hwc, 1, wc, c)
    t1 = torch.randn(c, h, w).cuda()  # stride (hw, w, 1)
    out = scripted_add(t0, t1)  # stride (hwc, 1, wc, c) preserving memory format of t0
```
3. different rank & compatible permutation with broadcasting
```
    t0 = torch.randn(b, h, w, c).cuda().permute([0, 3, 1, 2])  # stride (hwc, 1, wc, c)
    t1 = torch.randn(c).cuda().unsqueeze(-1).unsqueeze(-1)  # stride (1, 1, 1)
    out = scripted_add(t0, t1)  # stride (hwc, 1, wc, c) preserving memory format of t0
```
4. different rank & in-compatible permutation
```
    t0 = torch.randn(b, h, w, c).cuda().permute([0, 3, 1, 2])  # stride (hwc, 1, wc, c)
    t1 = torch.randn(h, w).cuda()  # stride (w, 1)
    jit_out = scripted_add(t0, t1)  # stride (hwc, 1, wc, c)  # stride (hwc, wc, c, 1)  # nvfuser outputs contiguous tensor
    eager_out = eager_add(t0, t1)  # stride (hwc, 1, wc, c)  # stride (hwc, 1, wc, c)  # TI preserves memory format of LHS operand
```
5. different rank & in-compatible permutation
```
    t0 = torch.randn(c, h, w).cuda()  # stride (hw, w, 1)
    t1 = torch.randn(b, h, w, c).cuda().permute([0, 3, 1, 2])  # stride (hwc, 1, wc, c)
    jit_out = scripted_add(t0, t1)  # stride (hwc, 1, wc, c)  # stride (hwc, 1, wc, c)  # nvfuser preserves memory format of highest rank tensors
    eager_out = eager_add(t0, t1)  # stride (hwc, 1, wc, c)  # stride (hwc, hw, w, 1)  # TensorIterator preserves memory format of LHS operand
```

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 1]
```diff
@@ -157,20 +163,106 @@ struct MemoryFormat {
...
       permutation_ = 0;
     }
   }
 
+  // returns the stride_order for given MemoryFormat permutation_
+  // see details for encoding in [ Note: MemoryFormat and Stride Order ]
+  std::vector<int> toStrideOrder() const {
+    std::vector<int> stride_order;
+    // be generous with reserved space
+    stride_order.reserve(10);
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: 10 is max dims supported right? Can you just make it a named constant?

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 2]
```diff
@@ -157,20 +163,106 @@ struct MemoryFormat {
...
     // special case permutation_ to reflect non-permuted tensor
-    if (!has_permutation_) {
+    if (!has_permutation) {
       permutation_ = 0;
     }
   }
 
+  // returns the stride_order for given MemoryFormat permutation_
+  // see details for encoding in [ Note: MemoryFormat and Stride Order ]
+  std::vector<int> toStrideOrder() const {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/parser.cpp:193]: This is from encoding to stride order?

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 3]
```diff
@@ -157,20 +163,106 @@ struct MemoryFormat {
...
+        }
+      } else if (cur_rank > rank) {
+        // shrink permutation to lower rank. We can simply discard higher rank
+        // stride order when they are not permuted to lower rank bit, because in
+        // those instance we can't obey broadcasting semantics while preserving
+        // permutation. We check for stride order and ensure that the lower
+        // `rank` bits are all <= rank. Afterwards, we update stride_order by
+        // decrement each entry by rank_diff to reflect correct stride order.
+        //
+        // e.g. for rank 4 channels last {1, 3, 2, 0}:
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/parser.cpp:257]: Is it safe for {3, 0, 2, 1} to be shrunk to {0, 2, 1}? or can it only be done from the right hand side. I'm missing what the syntax exactly means, is the integers the original position, and their current position is the new position or the reverse?

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 4]
```diff
@@ -138,17 +138,23 @@ struct MemoryFormat {
...
+  // [ Note: MemoryFormat and Stride Order ]
   // stride_order is extracted from
   //     `TensorType::stride_properties()::stride_index_`, it describes the
   // index of axes from fastest to slowest.
   // Look at comment for c10::Stride in aten/src/ATen/core/jit_type.h
   // e.g. for rank 4 non-permuted tensor, stride_order would be {3, 2, 1, 0}
+  //      it will be encoded as permutation_ = 3210
   //      for rank 4 channels last tensor, stride_order would be {1, 3, 2, 0}
+  //      it will be encoded as permutation_ = 1320
+  //      for rank 4 channels last tensor, stride_order would be {0, 3, 2, 1}
```
Comment (kevinstephano) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: Is there something omitted in this comment to explain what you are showing?

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 1 above)
Comment (kevinstephano) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: Nit? Perhaps move the reserve after you do the permutation check?

**Review by csarofeen — APPROVED** [review]
I think the logic makes sense to me, just had one real question.

**Review by kevinstephano — APPROVED** [review]
I think I just saw a couple of nitpicks.

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 1 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp]: I think we also have other places that set a lower standard (8). I'm gonna merge all somewhere

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 2 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp:193]: Yes. Will add another line in comment

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 4 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp]: I think that's indicating the comment is not clear enough~ Let me take another lap there.

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 3 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp:257]: No that's not safe... The check should be `is it safe for {3, 0, 2, 1} to be shrunk to rank 3`? and the answer is no

The second to the fasted dimension in the rank 4 tensor is axis-0. Which doesn't exist for the rank 3 tensor. So if we permute the rank 4 tensor (NCHW) into contiguous format, we'll have a new tensor (CHNW) <- notice `N` there, we can't have that dimension in the lower part.

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp:257]: One more comment is that, the permutation of the lower rank tensor doesn't matter, cause we can always permute it however we want.

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp:257]: In the words I have in the comment,
{3, 0, 2, 1} shrinking to rank 3 will require any@{3, 0, 2} >= (4-3); which is not true for `0` in the set.

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 4 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp]: Oh, wait... This doesn't look right.......

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 3 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/parser.cpp:257]: just reposting:
```
I'm missing what the syntax exactly means, is the integers the original position, and their current position is the new position or the reverse?
```

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp:257]: Oops. `stride order` is the same as `TensorType::stride_properties()::stride_index_`, it describes the index of axes from fastest to slowest.

so for a 4d tensor, if we have `stride_order = {x0, x1, x2, x3}`
The i-th fastest dimension would be `stride_order[i]`.
example 0: {3, 2, 1, 0} would mean the fastest dimension is axis-3. the next one would be 2, e.t.c.. So you'll see it's a contiguous tensor.
example 1: {1, 3, 2, 0} would mean the fastest dimension is axis-1. the next one would be 3, and then 2, and then 0. So this is a channels last tensor (NCHW).

I'll put this in the `[ Note: MemoryFormat and Stride Order]` and put a reference here.

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 1 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp]: Just a note, I'll do this in a separate PR to keep things simpler.

**Pull request merged**